library/psa_crypto.c

```diff
@@ -2992,7 +2986,12 @@ static psa_status_t psa_cipher_setup( psa_cipher_operation_t *operation,
         operation->iv_size = PSA_BLOCK_CIPHER_BLOCK_SIZE( slot->type );
     }
 
-    return( PSA_SUCCESS );
+exit:
+    if( status == 0 )
+        status = mbedtls_to_psa_error( ret );
+    if( status != 0 )
+        psa_cipher_abort( operation );
+    return( status );
```

Comment:
#68 suspected other setup functions might have issues with leaving around partially initialized contexts. Have our tests proven that they don't? It looks like the tests we've added in this PR just make sure the context is re-usable, not that we don't leave a partially initialized context around if setup fails.

Reply:
some things to look at: 
psa_cmac_setup - in this function if mbedtls_cipher_setup succeeds and mbedtls_cipher_cmac_starts fails
psa_generator_hkdf_setup - if psa_hmac_setup_internal succeeds but subsequesnt call to psa_hash_update fails.

Reply:
The tests `hash_setup`, `mac_setup` and `cipher_setup` check that the context is reusable. That's the only black-box way of testing that the context is properly reinitialized. There's no black-box way to test that a context isn't “partially initialized”: it's either sufficiently pristine for setup or not.

I don't think there's a way to test the case of `psa_cmac_setup` where `mbedtls_cipher_setup` succeeds and `mbedtls_cipher_cmac_starts` fails at the moment. An invalid key size is detected earlier, and an unsupported algorithm for CMAC can't be reached until we add support for Camellia and ARIA to `psa_crypto` (and at that point we may well allow them with CMAC).

I've deliberately not touched KDF code because the KDF API has upcoming changes and these bad-case tests will need to change accordingly.